http/src/v0/block.rs

```diff
@@ -98,31 +102,169 @@ pub fn put<T: IpfsTypes>(
+    DuplicateField(Cow<'a, str>),
+    MissingArg,
+    InvalidNumber(Cow<'a, str>, Cow<'a, str>),
+    InvalidBoolean(Cow<'a, str>, Cow<'a, str>),
+}
+
+impl<'a> fmt::Display for ParseError<'a> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        use ParseError::*;
+        match *self {
+            DuplicateField(ref s) => write!(fmt, "field {:?} was duplicated", *s),
+            MissingArg => write!(fmt, "required field \"arg\" missing"),
+            InvalidNumber(ref k, ref v) => write!(fmt, "field {:?} invalid number: {:?}", *k, *v),
+            InvalidBoolean(ref k, ref v) => write!(fmt, "field {:?} invalid boolean: {:?}", *k, *v),
+        }
+    }
+}
+
+impl<'a> TryFrom<&'a str> for RmOptions {
+    type Error = ParseError<'a>;
+
+    fn try_from(q: &'a str) -> Result<Self, Self::Error> {
+        use ParseError::*;
+
+        let parse = url::form_urlencoded::parse(q.as_bytes());
+
+        let mut args = Vec::new();
+        let mut force = None;
+        let mut quiet = None;
+
+        for (key, value) in parse {
+            let target = match &*key {
+                "arg" => {
+                    args.push(value.into_owned());
+                    continue;
+                }
+                "force" => &mut force,
+                "quiet" => &mut quiet,
+                _ => {
+                    // ignore unknown fields
+                    continue;
+                }
+            };
+
+            // common bool field handling
+            if target.is_none() {
+                match value.parse::<bool>() {
+                    Ok(value) => *target = Some(value),
+                    Err(_) => return Err(InvalidBoolean(key, value)),
+                }
+            } else {
+                return Err(DuplicateField(key));
+            }
+        }
+
+        Ok(RmOptions {
+            args,
+            force: force.unwrap_or(false),
+            quiet: force.unwrap_or(false),
+        })
+    }
+}
+
+use futures::stream::FuturesOrdered;
+
+async fn rm_query<T: IpfsTypes>(
+    ipfs: Ipfs<T>,
+    options: RmOptions,
+) -> Result<impl Reply, Rejection> {
+    println!("{:?}", options);
+
+    // let cids: Vec<Cid> = options
+    //     .args
+    //     .into_iter()
+    //     .map(|hash| Cid::from_str(&hash).map_err(StringError::from).unwrap())
+    //     .collect();
+
+    let futs: FuturesOrdered<_> = options
+        .args
+        .into_iter()
+        .map(|hash| Cid::from_str(&hash).map_err(StringError::from).unwrap())
+        .map(|cid| ipfs.remove_block(cid))
+        .collect();
+
+    let st = futs
+        .collect::<Vec<Result<Cid, Error>>>()
+        .await
+        .into_iter()
+        .map(|result| match result {
+            Ok(cid) => RmResponse {
+                hash: cid.to_string(),
+                error: "".to_string(),
+            },
+            Err(e) => {
+                RmResponse {
+                    // FIXME: How to get the `cid` here, though?
```

Comment:
you'll need to carry it next to the future on line [R232](https://github.com/ipfs-rust/rust-ipfs/pull/153/files#diff-ee377a5ddf5a7c29cad06ac34d1512edR232) like `.map(|cid| (cid.clone(), ipfs.remove_block(cid))` and fight a few compiler errors.

Reply:
I can't put .map_err after remove_block, which is async

Reply:
Not sure why? https://docs.rs/futures/0.3.4/futures/future/trait.TryFutureExt.html#method.map_err